crypto/compactcert/common.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package compactcert
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/big"
+
+	"github.com/algorand/go-algorand/crypto"
+)
+
+// hashCoin returns a number in [0, signedWeight) with a nearly uniform
+// distribution, "randomized" by all of the supplied arguments.
+func hashCoin(j uint64, sigcom crypto.Digest, signedWeight uint64) uint64 {
+	hashinput := make([]byte, 16+crypto.DigestSize)
+	binary.LittleEndian.PutUint64(hashinput[0:], j)
+	binary.LittleEndian.PutUint64(hashinput[8:], signedWeight)
+	copy(hashinput[16:], sigcom[:])
+	h := crypto.Hash(hashinput)
+
+	i := &big.Int{}
+	i.SetBytes(h[:])
+
+	w := &big.Int{}
+	w.SetUint64(signedWeight)
+
+	res := &big.Int{}
+	res.Mod(i, w)
+	return res.Uint64()
+}
+
+// numReveals computes the number of reveals necessary to achieve the desired
+// security parameters.  See section 8 of the ``Compact Certificates''
+// document for the analysis.
+//
+// numReveals is the smallest number that satisfies
+//
+// 2^-k >= 2^q * (provenWeight / signedWeight) ^ numReveals
+//
+// which is equivalent to the following, which avoids any floating-point math:
+//
+// signedWeight ^ numReveals >= 2^(k+q) * provenWeight ^ numReveals
+//
+// We use bigFloat for this computation, which is deterministic but adds
+// an error factor (rounding-down) of at most 1-2^-31 at each multiplication step.
+// The second-order (multiplied) error factors are negligible, but the additive
+// error factors add up to at most a multiplicative factor of 1-numReveals*2^-31.
```

Comment:
I think it's not quite as good as this comment, but still good enough. I wrote up a careful analysis here https://github.com/algorand/compactcert/blob/master/compactcert.pdf section 8.1 (p.6). In particular, the analysis (method 2) shows a cheap way to provably avoid underestimating: subtract 2 from mantissa of signedWeight after its canonicalized.

Reply:
Thanks for the careful analysis of this!

The error bounds on the security parameter seem small enough that it doesn't seem like method 1 vs method 2 matters much.  Do you think we should implement the `mantissa-2` trick?

Reply:
On the one hand, it's clear that what we are doing is fine. On the other hand, the argument for why it's fine is a bit subtle, and subtle security arguments are best avoided if possible. Just as a matter of simplicity. I am not a huge fan of my mantissa -2 solution, because it also involves a subtle argument. What about the following  simpler solution: add a boolean "roundUp" to setu64 and mul and set it to "false" for lhs / signedweight, and to "true" for rhs / provenweight. round Up will add 1 to the mantissa. The argument for why this is ok is much simpler: we round down on the side we are trying to prove is bigger, and round up on the side we are trying to prove is smaller, so we can't possible prove a false statement.

Reply:
Good idea; explicit rounding seems much clearer.

Pushed an implementation of explicit rounding, with compile-time enforcement that we never mix rounding modes (through separate types for rounded-up and rounded-down values).

Reply:
Looks good! I like the type-enforced limitations. One suggestion: do you want to enforce it on ge? 
` func (a *bigFloatDn) ge(b *bigFloatUp)`

Reply:
That seems like a nice idea.  Pushed it.